### ISSUE TITLE
Using z_injection_plane to "freeze" the beam for a given distance in a non-boosted-frame simulation

### ISSUE THREAD

**Issue by abonatto** [fbpic/fbpic#385]
Dear Developers,

I was looking for a way to keep the beam "frozen" for a given propagation distance (or time), and I thought z_injection_plane could work. By reading its description,

"z_injection_plane (float (in meters) or None) – When z_injection_plane is not None, then particles have a ballistic motion for z<z_injection_plane. This is sometimes useful in boosted-frame simulations. z_injection_plane is always given in the lab frame."

it seemed to me that it could be used for non-boosted-frame simulations. However, after trying, I saw that this is not the case: 

`AttributeError: 'BallisticBeforePlane' object has no attribute 'inv_gamma_boost'`

Does FBPIC has a way to prevent the beam to evolve for a given distance (or time)?

Regards,

Alex

**Comment by RemiLehe** [comment]
Thanks for reporting this issue!

In principle, `z_injection_plane` should work when running simulations without boosted frame, but you are correct that the current version of FBPIC contains a bug in that case.

I fixed it in PR #386. Once this PR is merged, you could test the bug fix by installing the `dev` version of FBPIC:
```
pip uninstall fbpic  # Uninstall previously-installed version of FBPIC
git clone https://github.com/fbpic/fbpic.git
cd fbpic
python setup.py install
```

With the bug fix, this should allow you to do what you want. Note however that the beam evolves ballistically before the injection plane. This means that the beam momenta are frozen, but that the positions of the particles still evolve. Thus, if the beam has a non-zero emittance, its shape (e.g. transverse RMS size) will still evolve, even before the injection plane. 
Is that acceptable for your use case?

**Comment by abonatto** [comment]
Dear Remi,

Thank you for your support, and for your explanation about the ballistic evolution of the beam. Indeed, I am looking for a way to prevent the transverse RMS size evolution while injecting a beam with non-zero emittance  into the plasma. Do you have any suggestion? Would it be a better approach to set the focus position (zf) to be equal to the beam longitudinal length (6*sigma_z, for example), in such a way the it will be at focus right after fully entering the plasma?

Regards,

Alex

**Comment by RemiLehe** [comment]
Hey, sorry for the late reply (I was away from the office for a few days)!

It all depends exactly what your needs are. It sounds like you would like the beam to reach focus at a given position (say `z=L0`) inside the plasma, but that you are (rightfully) worried that the forces felt by the beam inside the plasma might displace the focal position of the beam from the one you originally intended.

- If the evolution of the beam before `z=L0` is unimportant to you, then my advice would be to use `z_injection_plane=L0` (which is now fixed in the `dev` branch of FBPIC), `zf=L0` and `tf=(L0-z0)/c`, where `z0` is the centroid position of the beam at `t=0`. This will force the beam to reach focus at `z=L0`, but the RMS size of the beam will evolve before reaching `z=L0`.

- If it is important to you that the beam keeps the exact same RMS size all along its evolution before reaching `z=L0`, then one solution would be to initialize the beam with *zero* emittance, and then set its emittance to a non-zero value after `z=L0`. If you are interested in this solution, please let me know, as I could then post some more detailed instructions. (But it does get slightly more complicated and more "hacky".)

**Comment by abonatto** [comment]
Hi Remi,

No worries! Thanks again for your amazing support! I am definitely interested in the second option: could you please post the instructions?

Cheers,

Alex

**Comment by RemiLehe** [comment]
Sure, below is a "minimal" example (it does not have the plasma for instance, but that can be easily added). As mentioned previously, the key idea is to set the initial emittance to 0, but then to change the emittance to the desired value, by modifying the momenta of the beam directly. 
Let me know if you have any question:

```
import numpy as np
from scipy.constants import c, e, m_e
from fbpic.main import Simulation
from fbpic.lpa_utils.bunch import add_particle_bunch_gaussian
from fbpic.openpmd_diag import FieldDiagnostic, ParticleDiagnostic

# The simulation box
Nz = 400         # Number of gridpoints along z
zmax = 10.e-6    # Right end of the simulation box (meters)
zmin = -10.e-6   # Left end of the simulation box (meters)
Nr = 50          # Number of gridpoints along r
rmax = 20.e-6    # Length of the box along r (meters)
Nm = 1           # Number of modes used

# The simulation timestep
dt = (zmax-zmin)/Nz/c   # Timestep (seconds)

# The electron beam
L0 = 20.e-6      # Position at which the beam should be "unfreezed"
z0 = 5.e-6       # Initial centroid of the beam
sig_z = 2.e-6    # RMS size
sig_r = 2.e-6    # RMS size
n_emit = 1.e-6   # Normalized emittance in mm.mrad
gamma0 = 1000.
sig_gamma = 10.
Qtot = 10.e-12   # Charge in Coulomb

# Initialize the simulation object
sim = Simulation( Nz, zmax, Nr, rmax, Nm, dt, zmin=zmin, boundaries='open')
sim.set_moving_window( v=c )

# Create the beam *with zero emittance initially*
beam = add_particle_bunch_gaussian( sim, q=-e, m=m_e,
        sig_r=sig_r, sig_z=sig_z, n_emit=0, # initial emittance is 0
        gamma0=gamma0, sig_gamma=sig_gamma,
        n_physical_particles=Qtot/e, n_macroparticles=1000,
        zf=L0, tf=(L0-z0)/c, z_injection_plane=L0 )

# Add diagnostics
sim.diags = [ FieldDiagnostic( 50, sim.fld, comm=sim.comm ),
          ParticleDiagnostic( 50, {"beam": beam}, comm=sim.comm ) ]

# Let the beam propagate until it reaches the position L0
# (The beam evolves in a "frozen" way before that position,
# because of `z_injection_plane`)
t_frozen_propagation = (L0-z0)/c
N_steps_frozen = int( t_frozen_propagation / dt )
sim.step( N_steps_frozen )

# Now, set the emittance of the beam to the desired value,
# by directly changing the momenta of the beam
sig_u = n_emit/sig_r
beam.ux[:] = sig_u * np.random.normal( size=beam.Ntot )
beam.uy[:] = sig_u * np.random.normal( size=beam.Ntot )
beam.inv_gamma[:] = 1./np.sqrt( 1. + beam.ux**2 + beam.uy**2 + beam.uz**2 )

# Continue the simulation for as long as you wish
sim.step( 200 )
```

**Comment by abonatto** [comment]
Dear Remi,

Thank you very much for your detailed explanation. I will try it and let you know the outcome.

Cheers.

Alex

**Comment by RemiLehe** [comment]
@abonatto Btw, the above-mentioned bug-fix (i.e. fixing the ballistic injection in lab-frame simulations) is now included in the latest release of `fbpic` (0.13.1). Therefore, you do not need to use the `dev` version anymore, and you can use the official release instead:
```
pip uninstall fbpic # Uninstall the dev version
pip install fbpic # Installs the latest release
```

**Comment by abonatto** [comment]
Thanks @RemiLehe!
